### PR TITLE
fix: deprecate properties in app.getAppMetrics()

### PR DESCRIPTION
**Pull request by codebytere**
Related to https://github.com/electron/electron/issues/13383.

This PR deprecates two of the properties returned by `app.getAppMetrics()`, `sharedBytes` and `privateBytes` which are returned from chromium methods removed in [this PR](https://chromium-review.googlesource.com/c/chromium/src/+/969089). 

It also adds a new method to the deprecate module for removing properties of objects (`deprecate.removeFunction`), as well as a test for that method 🔧 

/cc @alexeykuzmin 

##### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] commit messages or PR title follow semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: docs/api/breaking-changes.md [hunk 1]
```diff
@@ -46,6 +46,11 @@ The following list includes the breaking API changes planned for Electron 3.0.
 app.getAppMemoryInfo()
 // Replace with
 app.getAppMetrics()
+
+// Deprecated
+const metrics = app.getAppMetrics()
+const privateBytes = metrics['privateBytes'] // deprecated property
+const sharedBytes = metrics['sharedBytes'] // deprecated property
```
Comment (alexeykuzmin) [inline docs/api/breaking-changes.md]: Maybe simple `metrics.sharedBytes` would look slightly better?

File: lib/browser/api/app.js [hunk 2]
```diff
@@ -58,6 +58,13 @@ Object.assign(app, {
   }
 })
 
+const nativeFn = app.getAppMetrics()
```
Comment (alexeykuzmin) [inline lib/browser/api/app.js]: I guess that you meant to save a reference to a function, and not to call it here.

File: lib/browser/api/app.js [hunk 3]
```diff
@@ -58,6 +58,13 @@ Object.assign(app, {
...
+app.getAppMetrics = () => {
+  deprecate.removeProperty(nativeFn, 'privateBytes')
+  deprecate.removeProperty(nativeFn, 'sharedBytes')
+  return nativeFn
```
Comment (alexeykuzmin) [inline lib/browser/api/app.js]: The function should be called here in the proper context:
```js
return nativeFn.call(app)
```

File: lib/browser/api/app.js (hunk 2 above)
Comment (codebytere) [reply lib/browser/api/app.js]: indeed i did!

File: spec/api-deprecations-spec.js [hunk 4]
```diff
@@ -76,6 +76,22 @@ describe('deprecations', () => {
...
 
+  it('deprecates a property of an object', () => {
+    let msg
+    deprecations.setHandler((m) => { msg = m })
+
+    const propertyName = 'itMustGo'
+    let value = 0
+
+    let o = { [propertyName]: value }
+    expect(o).to.have.a.property(propertyName).that.is.a('number')
```
Comment (alexeykuzmin) [inline spec/api-deprecations-spec.js]: I'm not sure if that `expect()` is really needed.
I just create this very object, and basically checks that the syntax is correct )

File: lib/browser/api/app.js [hunk 5]
```diff
@@ -58,6 +58,13 @@ Object.assign(app, {
   }
 })
 
+const nativeFn = app.getAppMetrics
+app.getAppMetrics = () => {
+  deprecate.removeProperty(nativeFn, 'privateBytes')
+  deprecate.removeProperty(nativeFn, 'sharedBytes')
```
Comment (alexeykuzmin) [inline lib/browser/api/app.js]: It looks slightly strange to me that the deprecation messages are printed for every method call, but I guess it's the best we can do here.

File: spec/api-deprecations-spec.js [hunk 6]
```diff
@@ -76,6 +76,22 @@ describe('deprecations', () => {
...
   })
 
+  it('deprecates a property of an object', () => {
+    let msg
+    deprecations.setHandler((m) => { msg = m })
+
+    const propertyName = 'itMustGo'
+    let value = 0
+
+    let o = { [propertyName]: value }
```
Comment (alexeykuzmin) [inline spec/api-deprecations-spec.js]: `const`

File: spec/api-deprecations-spec.js [hunk 7]
```diff
@@ -76,6 +76,22 @@ describe('deprecations', () => {
     expect(o).to.have.a.property(oldPropertyName).that.is.equal(value)
   })
 
+  it('deprecates a property of an object', () => {
+    let msg
+    deprecations.setHandler((m) => { msg = m })
+
+    const propertyName = 'itMustGo'
+    let value = 0
```
Comment (alexeykuzmin) [inline spec/api-deprecations-spec.js]: The var not really needed.

**Comment by eliranek1** [comment]
Can you please keep private and shared bytes? These are very critical for our reporting. You should be able to get that information straight from the OS by calling GetProcessMemoryInfo

**Comment by alexeykuzmin** [comment]
@eliranek1 They definitely won't be removed in Electron 3.0.x.
And we will try to preserve them in all further versions as well, see #13447.

**Pull request merged**